### PR TITLE
Add DeliveryLocation module

### LINKED ISSUE #54: Add Delivery Location Module
In this module i have added :
- functions that is related to the module
- module fields with checks and validators
- 0002_test_data to test the module
- register it to the admin
- added it to the project
- added tests to test the functionallity of the module

UI and user input and query to the client and suppliers will be done in a seperated PR and Issue.
- #53

### PR DESCRIPTION
**Pull request by thetaxas2468**
Adding the module deliverylocation with all its functions that is in the DFD for this feature with testing for migrations and for the data using pytest.
- closes #54

### CODE REVIEW

**Comment by thetaxas2468** [comment]
Dont look at the module i have some things to change like removing supplier and these things
all of these stuff will be changed when the supplier module is merged.
- In this PR at the moment i want the tests for migration and the module to be checked with the functions and DeliveryLocation module

File: deliverylocation/migrations/0001_initial.py
Comment (pazMenachem) [inline deliverylocation/migrations/0001_initial.py:1]: no need for this line

File: deliverylocation/models.py
Comment (pazMenachem) [inline deliverylocation/models.py:114]: why not try to full clean and then save? why save and then delete if fail?
same thing but more elegant as i see it.

File: deliverylocation/models.py
Comment (pazMenachem) [inline deliverylocation/models.py:125]: any validation available for the field? or it does it automatically?

**Review by pazMenachem — COMMENTED** [review]
other then the comments, Looks good! nice job!

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:114]: This is the proper way to write the full clean i guess

File: deliverylocation/migrations/0001_initial.py
Comment (thetaxas2468) [reply deliverylocation/migrations/0001_initial.py:1]: fixed

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:126]: fixed

File: deliverylocation/models.py
Comment (yftacherzog) [inline deliverylocation/models.py:7]: We have similar stuff on https://github.com/redhat-beyond/BuyTogether/pull/16/files. What's the reason to have it here as well?

Comment (thetaxas2468) [reply deliverylocation/models.py:7]: I know i have writted a comment at the begining,dont look at User SUpplier modules they are just helpers since the Supplier isnt merged yet,review other things and when all reviewd and fixed and supplier is merged i will rebase this PR on the upstream/main and update the paths and import the modules

**Comment by MeitarRizner** [comment]
What about the "delivery_location_id" field? I think you missed this field..

**Comment by pazMenachem** [comment]
> What about the "delivery_location_id" field? I think you missed this field..

if primary key is not mentioned, django creates one automatically👍

File: deliverylocation/models.py
Comment (Omeramsc) [reply deliverylocation/models.py:7]: If we should "not look at something" it usually points to this PR being a Work In Progress (WIP). 
As I mentioned earlier, we'd like to refrain from reviewing PRs which will have major changes in the forseen future.

Comment (thetaxas2468) [reply deliverylocation/models.py:7]: @Omeramsc I didnt understand what do you mean,can you explain more?

Comment (Omeramsc) [reply deliverylocation/models.py:7]: Sure: 
When you send a PR for review, you should respect the time of the reviewers (either mentors or random contributors on an open source project). usually when a PR is not complete and you just want someone to take a look and see you're going in the right direction, you'll mark it as a `work in progress`, a non-complete PR. 
If there's something we as reviewers should "not look at", it means this PR is indeed a work in progress.
If you'll mark it as one, I'll be able to review it in the right manner. otherwise I risk wasting time on reviewing something that will 99.9% change before this will be merged.

Comment (pazMenachem) [reply deliverylocation/models.py:7]: > Sure: When you send a PR for review, you should respect the time of the reviewers (either mentors or random contributors on an open source project). usually when a PR is not complete and you just want someone to take a look and see you're going in the right direction, you'll mark it as a `work in progress`, a non-complete PR. If there's something we as reviewers should "not look at", it means this PR is indeed a work in progress. If you'll mark it as one, I'll be able to review it in the right manner. otherwise I risk wasting time on reviewing something that will 99.9% change before this will be merged.

should we use comment in the code to indicate WIP or some how in the git? in other words, how to highlight the code section as WIP?

Comment (thetaxas2468) [reply deliverylocation/models.py:7]: But i want reviewers to review the code and fix all and the only change i need to do is rebase this on the upstreammain untill the supplier module is merged so i can rebase and delete the module from the dilvery location,functions tests will be the same, the only thing that will changed is deletion of the extra code for the supplier module and i metion that in the comments

Comment (Omeramsc) [reply deliverylocation/models.py:7]: @pazMenachem  It is not the part of code that is a WIP, but the whole PR. 
@thetaxas2468 So ask for reviewers anyway, but rebase is never "just rebase". things can change, and those things might have us review the PR again. 
so it is OK to ask for a review on a WIP or a PR that (heavily) depends on another, but you should mark it as so. It'll also make it easier to understand.

Comment (thetaxas2468) [reply deliverylocation/models.py:7]: How do i mark the whole PR as WIP and in the same time ask reviewers to review

Comment (Omeramsc) [reply deliverylocation/models.py:7]: Edit the PR- there should be an option to mark it as a WIP
but even if there isn't, just add `WIP: ` at the beginning. 
then tag whoever you want to review as a reviewer, and explain yourself.

**Comment by thetaxas2468** [comment]
[WIP] PR
Reviewers : @yftacherzog @Omeramsc @RawanAbuleil @pazMenachem @LiorSilberman @MeitarRizner

Things to review: Everything from functions to tests but the only thing to ignore is the Module User With Supplier(Becuase it is not merged so i need to do something like this in order to know what a supplier is in this module)

!When finish reviewing do not merge,rebase on merged supplier module is needed!

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:7]: Done

File: deliverylocation/models.py
Comment (yftacherzog) [inline deliverylocation/models.py:119]:
```suggestion
    def update_date(self, newdate):
```
As this is a method of the `DeliveryLocation` object, then I think you can drop the `delivery_location` from the name. It will make things look cleaner.

File: deliverylocation/models.py
Comment (yftacherzog) [inline deliverylocation/models.py:123]:
```suggestion
    def update_location(self, newlocation):
```

File: deliverylocation/models.py
Comment (yftacherzog) [inline deliverylocation/models.py:124]: Do we want to ignore illegal input here and just silently skip? Can we instead remove the `if` and call `full_clean` or something similar?

File: deliverylocation/models.py
Comment (yftacherzog) [inline deliverylocation/models.py:129]:
```suggestion
    def filter_by_supplier_and_location(supplier, specific_location):
```

File: deliverylocation/models.py
Comment (yftacherzog) [inline deliverylocation/models.py:133]:
```suggestion
    def filter_by_location(supplier):
```

File: deliverylocation/models.py
Comment (yftacherzog) [inline deliverylocation/models.py:141]: snake_case for args

File: deliverylocation/models.py
Comment (yftacherzog) [inline deliverylocation/models.py:145]: snake_case for args

File: deliverylocation/tests/test_deliverylocation.py
Comment (yftacherzog) [inline deliverylocation/tests/test_deliverylocation.py:9]:
```suggestion
test_data = [
    ('taxas', 'firstname', 'lastname', 'password', 'supplier', 'business name'),
    ('ed', 'edd', 'eddy', 'jawbreaker101', 'supplier', 'jawbreaker inc'),
]
```

File: deliverylocation/tests/test_deliverylocation.py
Comment (yftacherzog) [inline deliverylocation/tests/test_deliverylocation.py:55]: What's the incentive for doing it twice? If it changes for some reason, you have to change it twice.

File: deliverylocation/tests/test_deliverylocation.py
Comment (yftacherzog) [inline deliverylocation/tests/test_deliverylocation.py:65]: You're testing two different methods. Each should be in its own test case.

File: deliverylocation/tests/test_deliverylocation.py
Comment (yftacherzog) [reply deliverylocation/tests/test_deliverylocation.py:55]: I'd like to see as well a test in which the result is empty (same for other filters as well).

File: deliverylocation/tests/test_deliverylocation.py
Comment (yftacherzog) [inline deliverylocation/tests/test_deliverylocation.py:31]: I think it would have make sense to make this a fixture. Or just put it inside of the `deliverylocation0` fixture.

File: deliverylocation/admin.py
Comment (yftacherzog) [inline deliverylocation/admin.py:2]: I think the package name should be `delivery_location` rather than `deliverylocation`.

File: deliverylocation/apps.py
Comment (yftacherzog) [inline deliverylocation/apps.py:4]:
```suggestion
class DeliveryLocationConfig(AppConfig):
```

File: deliverylocation/tests/test_deliverylocation.py
Comment (yftacherzog) [inline deliverylocation/tests/test_deliverylocation.py:18]: I'd prefer it in snake_case. Same goes to everything in this file which is function or variable name.

File: deliverylocation/admin.py
Comment (thetaxas2468) [reply deliverylocation/admin.py:2]: Done!

File: deliverylocation/apps.py
Comment (thetaxas2468) [reply deliverylocation/apps.py:4]: Done

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:119]: Done

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:123]: Done

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:124]: newlocation is a string,if illegal input must be something like null which doesnt have len >=1

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:129]: Done

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:133]: Done

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:141]: fixed

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py:145]: fixed

File: deliverylocation/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply deliverylocation/tests/test_deliverylocation.py:9]: Fixed

File: deliverylocation/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply deliverylocation/tests/test_deliverylocation.py:31]: tried to do that but it makes it more complicated to understand what is being saved and deleted, and it fails some tests due the deletion and saving back

File: deliverylocation/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply deliverylocation/tests/test_deliverylocation.py:18]: Done

File: deliverylocation/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply deliverylocation/tests/test_deliverylocation.py:65]: Fixed

File: deliverylocation/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply deliverylocation/tests/test_deliverylocation.py:55]: Done

File: deliverylocation/models.py
Comment (yftacherzog) [reply deliverylocation/models.py]: `len(None)` will raise an exception. You can use instead:
```
if newlocation:
```
but still, I don't understand the logic. having an update function that fails silently looks risky -- or at least bad user experience (imagine you update a value and it stays the way it was without letting you know why).

File: delivery_location/models.py
Comment (yftacherzog) [inline delivery_location/models.py:141]: typo: dilvery -> delivery

File: delivery_location/tests/test_deliverylocation.py
Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:55]: This should be on its own test case.

File: delivery_location/tests/test_deliverylocation.py
Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:61]: You don't actually use `delivery_location1` here. Why do you need it added.

File: delivery_location/tests/test_deliverylocation.py
Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:9]: If it's a const then it needs to be in uppercase, but it's better to have that as a fixture. This way, if one of the tests accidentally modifies it, you get a fresh copy on the next test.

File: delivery_location/tests/test_deliverylocation.py
Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:64]: I suggest moving the second assertion to a different test case.

File: delivery_location/tests/test_deliverylocation.py
Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:74]: different test case

File: delivery_location/models.py
Comment (thetaxas2468) [reply delivery_location/models.py:141]: fixed

File: delivery_location/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py:9]: fixed

File: delivery_location/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py:55]: fixed

File: delivery_location/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py:61]: fixed

File: delivery_location/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py:64]: fixed

File: delivery_location/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py:74]: fixxed

File: deliverylocation/models.py
Comment (thetaxas2468) [reply deliverylocation/models.py]: Updated(Also the logic that the new location is bigger than zero going to be in frontend pages too)
i have added if len = 0 or newlocation is None validationerror will be raised with specific message,other than that it will succeed

File: delivery_location/models.py
Comment (yftacherzog) [inline delivery_location/models.py:126]:
```
In [76]: bool([])
Out[76]: False

In [77]: bool(None)
Out[77]: False

In [78]: bool("s")
Out[78]: True
```
You don't need the `elif`. If the location was 0, you'd end up on the first `if`

File: delivery_location/models.py
Comment (yftacherzog) [inline delivery_location/models.py:129]: You'd never get here.

File: delivery_location/tests/test_deliverylocation.py
Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:8]: now that it's a fixture, it needs to be in lowercase. Sorry if I misled you.

File: delivery_location/tests/test_deliverylocation.py
Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:57]: This should be on its own test case.

Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:57]: This should be in a test case of its own

File: delivery_location/tests/test_deliverylocation.py
Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:63]: Why do we need to add this one?

File: delivery_location/tests/test_deliverylocation.py
Comment (yftacherzog) [inline delivery_location/tests/test_deliverylocation.py:85]: I'm not sure whether we can tell for sure `filter_by_location` will always return the same order of locations. If not, this test can become flaky (test with unstable results). I'd suggest sorting the result:

```suggestion
        assert sorted([delivery_location0, delivery_location1]) == sorted(list(DeliveryLocation.filter_by_location(supplier0)))
```
But it might be that those objects don't have natural order according to which they can be sorted, so if this fails, you should be able to do something like this:

```suggestion
        assert sorted([delivery_location0, delivery_location1], key=str) == sorted(list(DeliveryLocation.filter_by_location(supplier0)), key=str)
```

Another option will be to check that both values are in the returned list + checking the length of the returned list. Something like:
```
output = list(DeliveryLocation.filter_by_location(supplier0))
locations = [delivery_location0, delivery_location1]
assert all(location in locations for location in output)
assert len(output) == len(locations)
```

File: delivery_location/models.py
Comment (thetaxas2468) [reply delivery_location/models.py:126]: You are right,fixed!

File: delivery_location/models.py
Comment (thetaxas2468) [reply delivery_location/models.py:129]: fixed

File: delivery_location/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py:8]: Its okay,its a new thing to learn
Fixed!

File: delivery_location/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py:57]: fixed

File: delivery_location/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py:63]: Its removed,its a duplicate
fixed

File: delivery_location/tests/test_deliverylocation.py
Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py:85]: You are completely right,i need to sort them in order to check that both of them identical,thanks!
fixed

Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py]: Or we can use assert sorted([delivery_location0, delivery_location1], key=lambda x:x.id) == sorted(list(DeliveryLocation.filter_by_location(supplier0)), key=x:x.id)
Which one do you prefer other than the str.

Comment (yftacherzog) [reply delivery_location/tests/test_deliverylocation.py]: Whatever you prefer.

Another option which might be even cleaner is to compare them as sets. The order in sets doesn't matter, but they include no duplicate instances.

If your function could have returned duplicate instances, then this would not have been a good use of sets, but in this case, I don't think you can get duplicate instances from your database, so I think that should be ok.

```
In [18]: set([1,2,3,3])
Out[18]: {1, 2, 3}

In [19]: [1,2,3] == [3,2,1]
Out[19]: False

In [20]: set([1,2,3]) == set([3,2,1])
Out[20]: True
```

So, if you call `set()` on both sides of this comparison, I think it will still be legitimate comparison and cleaner than the other approaches.

Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py]: Thanks i didnt know that comparision betwen sets doesnt matter but i didnt know that if the order is not the same with python then the same case happends, thanks new things to use in the future,fixed to be set since it is the right case.
Fixed!

Comment (yftacherzog) [reply delivery_location/tests/test_deliverylocation.py]: Cool. Sets are convenient for making stable tests. Just remember not to use them if you care about repetitions or order of elements.

Comment (thetaxas2468) [reply delivery_location/tests/test_deliverylocation.py]: Understood!

**Review by yftacherzog — APPROVED** [review]

File: delivery_location/models.py
Comment (Omeramsc) [inline delivery_location/models.py:148]: I see here a bunch of meaningless wrapper methods... and a bunch of meaningless tests for all of them. those are Django inner methods, we don't need to rewrite or test them - The people working on Django have already done it. Please don't waste time on it.

**Review by Omeramsc — CHANGES_REQUESTED** [review]
- Is this still a WIP? if not, please modify the title.
- Fix conflict
- Fix my comments

File: delivery_location/models.py
Comment (thetaxas2468) [reply delivery_location/models.py:148]: Regular delete method does not raise any doesnotexist exception it only raises protected exception but in our case we didnt use any protected fields all of them cascade,since i want to get an exception if specific id doesnt exist so i use the get method that raises Object(DilverLocation)DoesNotExist exception

Comment (yftacherzog) [reply delivery_location/models.py:148]: There's still a race condition here, BTW. It's still possible that the object will be there by the time you fetch it using get and will not be there by the time the delete member method is called.

I agree that having those wrappers is a bit redundant and generates additional maintenance effort for which I still don't see the benefit.

Comment (thetaxas2468) [reply delivery_location/models.py:148]: When a supplier deletes a dilvery location i want the supplier to know that he deleting something that exist other wise he gets a meesage like exception,regular delete method does not rise exception even if the object we are trying to delete is not there in the database,so i want something that can warn us that we are deleting something that does not exist.

Comment (yftacherzog) [reply delivery_location/models.py:148]: I see the point, and this is why I was ok with that on other modules as well.

Having said that, this logic still contains the race condition I mentioned above, which practically brings us to square one, so it's not a deal-breaker for me, but it's something that you should be aware of if you try to further build logic on top of this later on (I can't see at the moment how it might get you troubles -- but keep it in mind).

Comment (thetaxas2468) [reply delivery_location/models.py:148]: Are you talking about the situation when we have two fetches one for get the object and one for delete the object,so when the delete method called it stops at the get method and the second fetch get the object that has been deleted(but not deleted from the database yet)?

Comment (yftacherzog) [reply delivery_location/models.py:148]: As the code is written now, you're ultimately using the member's `delete()` method, you're just making sure it exists in the database by fetching the object from the database before deleting it.

What I'm saying is that it might be that the object will be deleted (by someone else) between the time you fetched the object (`DeliveryLocation.objects.get(id=a_id)`) and the time you call its `delete()` method (`.delete()`).

Again, I'm not seeing any problem with that at the moment, but just keep this race condition in mind.

Comment (thetaxas2468) [reply delivery_location/models.py:53]: Yes since it is deleted by someone else and i am trying to delete it by the time it is deleted it is fine because no exception will be raised(we will be doing that only the supplier that has the deliverylocation can actually delete his stuff in the future)

**Comment by thetaxas2468** [comment]
> * Is this still a WIP? if not, please modify the title.
> * Fix conflict
> * Fix my comments

- Title modified
- conflict has been solved
- your comment is discussed and fixed

**Comment by thetaxas2468** [comment]
> Dont look at the module i have some things to change like removing supplier and these things all of these stuff will be changed when the supplier module is merged.
> 
> * In this PR at the moment i want the tests for migration and the module to be checked with the functions and DeliveryLocation module

fixed!

File: delivery_location/models.py
Comment (yftacherzog) [reply delivery_location/models.py:53]: I hijacked the thread a bit, but I think @Omeramsc's comment was about some of the other static methods as well.

**Review by yftacherzog — APPROVED** [review]

File: delivery_location/models.py
Comment (thetaxas2468) [reply delivery_location/models.py:53]: What other static methods,all has been done in the same logic with tests

File: delivery_location/models.py
Comment (pazMenachem) [inline delivery_location/models.py:45]: omer meant that those two are irrelevant, no point warping those methods, the rest raises an error so i can understand.
technically if you create a class method ull need to test it, like omer said no point of doing so since they do the same as built in tested methods.

Comment (thetaxas2468) [reply delivery_location/models.py:45]: These functions added so the logic of using the deliverylocation is known and in order to use that logic you just simply use the functions that is created and if there is functions that is needed in the future i will update the deliverylocation,Instead of writing the DeliveryLocation.objects.all() as a response of an http get request for example we use get_all_delivery_locations that is more understandable for example

Comment (pazMenachem) [reply delivery_location/models.py:45]: i still dont see the reason, get all delivery... is a static method, writing DeliveryLocation.get_all_delivery_location() is same as writing DeliveryLocation.objects.all(), both static and no extra feature was added.
if you believe theres a difference that i dont see, then ok🤷‍♂️

Comment (thetaxas2468) [reply delivery_location/models.py:45]: Lets see you are a client that wants to check a code, which one is more understandable
object.all
or
get_all_delivery_locations

the same thing happend when we write code and trying to debug something if something goes wrong

File: delivery_location/models.py
Comment (Omeramsc) [reply delivery_location/models.py:53]: I still don't see how this method is needed. where is the change in the exception coming from? all we're doing is calling `get` and `delete`, we can do the exact same thing without this function, yielding the same result. there is no special exception handling or a use of full clean here.

**Comment by Omeramsc** [comment]
Can we take this PR out of `in progress`?

File: delivery_location/models.py
Comment (thetaxas2468) [reply delivery_location/models.py:53]: We are using two functions that are written in one,so in order to use thos two functions we call the static function that has a meaning of what we are doing instead of just code and in the future this function can be written differently in order to write that abstract two functions in views handling we just write the static method that is already writted with more logic

**Comment by thetaxas2468** [comment]
> Can we take this PR out of `in progress`?

Yes this pr is ready to review or even ready to be merged

File: delivery_location/models.py
Comment (Omeramsc) [reply delivery_location/models.py:53]: I don't fully agree with this solution, but I understand what you were going for. I'll approve it.

**Review by Omeramsc — APPROVED** [review]
LGTM
merged

**Pull request merged**
